### PR TITLE
[SPARK-28885][SQL] Follow ANSI store assignment rules in table insertion by default

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

When inserting a value into a column with the different data type, Spark performs type coercion. Currently, we support 3 policies for the store assignment rules: ANSI, legacy and strict, which can be set via the option "spark.sql.storeAssignmentPolicy":
1. ANSI: Spark performs the type coercion as per ANSI SQL. In practice, the behavior is mostly the same as PostgreSQL. It disallows certain unreasonable type conversions such as converting `string` to `int` and `double` to `boolean`. It will throw a runtime exception if the value is out-of-range(overflow).
2. Legacy: Spark allows the type coercion as long as it is a valid `Cast`, which is very loose. E.g., converting either `string` to `int` or `double` to `boolean` is allowed. It is the current behavior in Spark 2.x for compatibility with Hive. When inserting an out-of-range value to a integral field, the low-order bits of the value is inserted(the same as Java/Scala numeric type casting). For example, if 257 is inserted to a field of Byte type, the result is 1.
3. Strict: Spark doesn't allow any possible precision loss or data truncation in store assignment, e.g., converting either `double` to `int` or `decimal` to `double` is allowed. The rules are originally for Dataset encoder. As far as I know, no mainstream DBMS is using this policy by default.

Currently, the V1 data source uses "Legacy" policy by default, while V2 uses "Strict". This proposal is to use "ANSI" policy by default for both V1 and V2 in Spark 3.0.

### Why are the changes needed?

Following the ANSI SQL standard is most reasonable among the 3 policies. 

### Does this PR introduce any user-facing change?

Yes. 
The default store assignment policy is ANSI for both V1 and V2 data sources.

### How was this patch tested?

Unit test

### CODE REVIEW

**Comment by cloud-fan** [comment]
let's also add a migration guide for DS v1.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcQuerySuite.scala
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcQuerySuite.scala]: This query writes array not string.

Comment (gengliangwang) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcQuerySuite.scala]:
```
> spark.sql("select array()")
res0: org.apache.spark.sql.DataFrame = [array(): array<string>]
```

I think it makes more sense to make the type of empty array as `array<null>`.
How about create a jira for this and disable the test case for now?

Comment (gengliangwang) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/HiveOrcQuerySuite.scala]: Created https://issues.apache.org/jira/browse/SPARK-29462 for this.

**Comment by gengliangwang** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala]: I think we can write null to any nullable column even if it's strict policy. We can have a followup PR to discuss it further.

**Comment by cloud-fan** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala]: Okay.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. This is a big change according to the vote. 
Thank you, @gengliangwang and @cloud-fan .
I'll merge this to master to unblock the other PRs. This opened up several follow-up issues. Also, we can adjust the rule on `strict` policy later.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: @gengliangwang Sorry, I just realized recently that my changed are not tested by `date.sql` and `timestamp.sql` any more when I run: `build/sbt "sql/test-only *SQLQueryTestSuite -- -z timestamp.sql"`. Did you disable them forever or are they executed in some way by jenkins?

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: How about just setting `storeAssignmentPolicy=LEGACY` for the PgSQL tests? They have a lots of INSERT queries having implicit casts from string literals to numeric values.

Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: @maropu I am not sure about that. 
For PgSQL, it disable inserting string values to numeric columns except for literals. Setting `storeAssignmentPolicy=LEGACY` for all the PgSQL tests seems inaccurate.

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: Ur, I see...

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: But, before this pr's been merged, we tested the PgSQL tests in the LEGACY mode? Is my understanding wrong? Personally, I think we need to explicitly file issues in jira if we have inaccurate tests in `pgSQL/`.

Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: How about changing the timestamp/date values from string literal to timestamp/date literal in those sql files, just as https://github.com/apache/spark/pull/26107/files#diff-431a4d1f056a06e853da8a60c46e9ca0R68
I am not sure about whether there is a guideline in porting the PgSQL test files. Such modifications are allowed, right?

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: > I am not sure about whether there is a guideline in porting the PgSQL test files. Such modifications are allowed, right?

Yea, I think that's ok.  cc: @wangyum @dongjoon-hyun @HyukjinKwon 
If you have no time to do, I'll will check next week.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: I'm +1 for @gengliangwang 's suggestion `enabling those tests with that modification`.

**Pull request closed without merging**
